Tools/clinic/clinic.py

```diff
@@ -5158,6 +5143,66 @@ def parse_cloned_function(self, names: FunctionNames, existing: str) -> None:
         (cls or module).functions.append(function)
         self.next(self.state_function_docstring)
 
+    @staticmethod
+    def generate_c_basename(full_name: str) -> str:
+        fields = full_name.split(".")
+        if fields[-1] == '__new__':
+            fields.pop()
+        return "_".join(fields)
+
+    def parse_declaration(
+        self, line: str
+    ) -> tuple[str, str, str | None, str | None]:
+        cloned = None
+        returns = None
+
+        def invalid_syntax(msg: str | None = None) -> NoReturn:
+            preamble = "Invalid syntax"
+            if msg:
+                preamble += f" ({msg})"
+            fail(f"{preamble}: {line!r}\n\n"
+                 "Allowed syntax:\n"
+                 "[module.[submodule.]][class.]func [as c_name] [-> return_annotation]\n\n"
+                 "Nested submodules are allowed.")
+
+        m = RE_FULLNAME.match(line)
+        assert m
+        full_name = m[1]
+        if not libclinic.is_legal_py_identifier(full_name):
+            fail(f"Illegal function name: {full_name!r}")
+        pos = m.end()
+
+        m = RE_C_BASENAME.match(line, pos)
+        if m:
+            if not m[1]:
+                fail(f"No C basename provided for {full_name!r} after 'as' keyword")
+            c_basename = m[1]
+            if not libclinic.is_legal_c_identifier(c_basename):
+                fail(f"Illegal C basename: {c_basename!r}")
+            pos = m.end()
+        else:
+            c_basename = self.generate_c_basename(full_name)
+
+        m = RE_CLONE.match(line, pos)
+        if m:
+            if not m[1]:
+                fail(f"No source function provided for {full_name!r} after '=' keyword")
+            cloned = m[1]
+            if not libclinic.is_legal_py_identifier(cloned):
+                fail(f"Illegal source function name: {cloned!r}")
+            pos = m.end()
+
+        m = RE_RETURNS.match(line, pos)
+        if m:
+            if cloned:
+                invalid_syntax()
+            if not m[1]:
+                fail(f"No return annotation provided for {full_name!r} after '->' keyword")
+            returns = m[1].strip()
```

Comment:
I forget to add this:
```suggestion
            returns = m[1].strip()
            pos = m.end()

        if end != len(line):
            invalid_syntax()
```

Reply:
Makes sense; that helps a lot. Added in 3c83c5d2211d9d92ba19016f17d8dd79b7e34fe9